buildhub2 doesn't run in local dev environment

Buildhub doesn't run in the local dev environment. I'm hitting a bunch of problems. This issue covers fixing things so we can run it locally.

First, the ui isn't proxying web requests. I think it's because it's set to proxy to localhost:8000 which doesn't exist in the ui container. Changing "proxy" to "http://web:8000" fixes that.

Second, I'm hitting this assertion error in the web container:

```
eb_1            | ERROR 2019-05-31 15:02:01,454 django.request Internal Server Error: /sockjs-node/357/mrp440zw/websocket
web_1            | Traceback (most recent call last):
web_1            |   File "/usr/local/lib/python3.7/site-packages/django/core/handlers/exception.py", line 34, in inner
web_1            |     response = get_response(request)
web_1            |   File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 126, in _get_response
web_1            |     response = self.process_exception_by_middleware(e, request)
web_1            |   File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 124, in _get_response
web_1            |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
web_1            |   File "/app/buildhub/urls.py", line 30, in serve
web_1            |     assert os.path.isdir(document_root), document_root
web_1            | AssertionError: /app/ui/build
web_1            | ERROR 2019-05-31 15:02:01,455 django.server "GET /sockjs-node/357/mrp440zw/websocket HTTP/1.1" 500 71239
```

That directory gets created when building the image, but because it's in `/app` and `/app` is volume mounted repo directory, there's no `/app/ui/build` in the repo and then it fails. We had this problem with Socorro as well and instead of building static assets in `/app`, we built them in another directory outside of `/app`.

Another way to deal with this is to have the run script build the static assets if they don't exist and then run the web. I think I may do that here.

--- a/.env-dist
+++ b/.env-dist
@@ -7,7 +7,7 @@ DJANGO_SECRET_KEY=dontusethisinprod
 # Only override this if you need something different from the default
 DJANGO_ALLOWED_HOSTS=web,localhost
 
-DATABASE_URL=postgresql://postgres@db/postgres
+DATABASE_URL=postgresql://postgres@db/buildhub2
 DJANGO_ES_URLS=http://elasticsearch:9200
 
 DJANGO_S3_BUCKET_URL=https://s3-us-west-2.amazonaws.com/buildhub-sqs-test

--- a/Makefile
+++ b/Makefile
@@ -10,6 +10,7 @@ default:
 	@echo "Welcome to the buildhub\n"
 	@echo "The list of commands for local development:\n"
 	@echo "  build            Builds the docker images for the docker-compose setup"
+	@echo "  setup            Initializes and sets up Postgres and Elasticsearch services"
 	@echo "  run              Runs the whole stack, served on http://localhost:8000/"
 	@echo "  gunicorn         Runs the whole stack using gunicorn on http://localhost:8000/"
 	@echo "  daemon           Start the SQS daemon"
@@ -38,6 +39,10 @@ build: .env
 	docker-compose build
 	touch .docker-build
 
+.PHONY: setup
+setup: .env
+	docker-compose run web /app/bin/setup-services.sh
+
 .PHONY: clean
 clean: .env stop
 	docker-compose rm -f
@@ -46,7 +51,6 @@ clean: .env stop
 
 .PHONY: migrate
 migrate: .env
-	# docker-compose run web python manage.py migrate --run-syncdb
 	docker-compose run web python manage.py migrate
 
 .PHONY: shell
@@ -61,7 +65,7 @@ currentshell: .env .docker-build
 
 .PHONY: psql
 psql: .env .docker-build
-	docker-compose run db psql -h db -U postgres
+	docker-compose run db psql --host db --username postgres
 
 .PHONY: stop
 stop: .env

--- a/bin/db.py
+++ b/bin/db.py
@@ -1,0 +1,78 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import argparse
+import os
+import sys
+from urllib.parse import urlparse
+
+import psycopg2
+from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
+
+
+DESCRIPTION = 'Create and delete the database set by DATABASE_URL.'
+
+EPILOG = 'Required DATABASE_URL to be set in the environment.'
+
+
+def create_database(dsn):
+    parsed = urlparse(dsn)
+    db_name = parsed.path[1:]
+    adjusted_dsn = dsn[:-(len(db_name) + 1)]
+
+    conn = psycopg2.connect(adjusted_dsn)
+    conn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+    cursor = conn.cursor()
+    try:
+        cursor.execute('CREATE DATABASE %s' % db_name)
+        print('Created database "%s".' % db_name)
+    except psycopg2.ProgrammingError:
+        print('Database "%s" already exists.' % db_name)
+        return 1
+
+
+def drop_database(dsn):
+    parsed = urlparse(dsn)
+    db_name = parsed.path[1:]
+    adjusted_dsn = dsn[:-(len(db_name) + 1)]
+
+    conn = psycopg2.connect(adjusted_dsn)
+    conn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+    cursor = conn.cursor()
+    try:
+        cursor.execute('DROP DATABASE %s' % db_name)
+        print('Database "%s" dropped.' % db_name)
+    except psycopg2.ProgrammingError:
+        print('Database "%s" does not exist.' % db_name)
+        return 1
+
+
+def main(argv=None):
+    parser = argparse.ArgumentParser(
+        description=DESCRIPTION.strip(),
+        epilog=EPILOG.strip()
+    )
+    subparsers = parser.add_subparsers(dest='cmd')
+    subparsers.required = True
+    subparsers.add_parser('drop', help='drop existing database')
+    subparsers.add_parser('create', help='create database')
+
+    args = parser.parse_args()
+
+    try:
+        dsn = os.environ['DATABASE_URL']
+    except KeyError:
+        dsn = ''
+
+    if not dsn:
+        parser.error('DATABASE_URL is not set in environment')
+
+    if args.cmd == 'drop':
+        return drop_database(dsn)
+    elif args.cmd == 'create':
+        return create_database(dsn)
+
+
+if __name__ == '__main__':
+    sys.exit(main(sys.argv[1:]))

--- a/bin/run.sh
+++ b/bin/run.sh
@@ -42,7 +42,6 @@ case $1 in
     ${CMD_PREFIX_GUNICORN:-gunicorn} buildhub.wsgi:application -b 0.0.0.0:${PORT} --timeout ${GUNICORN_TIMEOUT} --workers ${GUNICORN_WORKERS} --access-logfile -
     ;;
   web-dev)
-    ${CMD_PREFIX_PYTHON:-python} manage.py migrate --noinput
     ${CMD_PREFIX_PYTHON:-python} manage.py runserver 0.0.0.0:${PORT}
     ;;
   blackfix)

--- a/bin/setup-services.sh
+++ b/bin/setup-services.sh
@@ -1,0 +1,17 @@
+#!/usr/bin/env bash
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+set -eo pipefail
+
+# Drops and recreates Postgres db and Elasticsearch indexes.
+
+# Set up db
+python bin/db.py drop || true
+python bin/db.py create
+python manage.py migrate --noinput
+
+# Delete and recreate Elasticsearch index
+python manage.py reindex-elasticsearch

--- a/buildhub/urls.py
+++ b/buildhub/urls.py
@@ -26,7 +26,6 @@ def serve(request, **kwargs):
     _, ext = os.path.splitext(request.path_info)
     if ext:
         return http.HttpResponseNotFound(request.path_info)
-    document_root = kwargs["document_root"]
 
     response = django_serve(request, "/index.html", **kwargs)
     if isinstance(response, http.FileResponse):

--- a/buildhub/urls.py
+++ b/buildhub/urls.py
@@ -27,7 +27,6 @@ def serve(request, **kwargs):
     if ext:
         return http.HttpResponseNotFound(request.path_info)
     document_root = kwargs["document_root"]
-    assert os.path.isdir(document_root), document_root
 
     response = django_serve(request, "/index.html", **kwargs)
     if isinstance(response, http.FileResponse):

--- a/docs/dev.rst
+++ b/docs/dev.rst
@@ -2,6 +2,43 @@
 Developer Documentation
 =======================
 
+Setting up a development environment
+====================================
+
+You can set up a development environment with:
+
+.. code-block:: shell
+
+   # Builds Docker containers
+   $ make build
+
+   # Drops and recreates Postgres db and Elasticseach index
+   $ make setup
+
+
+Buildhub2 has two parts: a webapp and a daemon.
+
+To run the webapp, do:
+
+.. code-block:: shell
+
+   # Runs web and ui and required services
+   $ make run
+
+This will start a server that is exposed on port ``8000`` so now you can
+reach ``http://localhost:8000`` with your browser or curl.
+
+It will also start the ``create-react-app`` dev server on port ``3000``. That's
+the main URL to use.
+
+To run the daemon that polls SQS for events and generates build information:
+
+.. code-block:: shell
+
+   $ make daemon
+
+This will run until an unexpected error happens or until you kill it with ``Ctrl-C``.
+
 Code
 ====
 
@@ -39,30 +76,6 @@ To run lint-based fixing tasks on Python and JS files:
 .. code-block:: shell
 
    $ make lintfix
-
-Local development
-=================
-
-You run everything in Docker with:
-
-.. code-block:: shell
-
-   $ make build  # only needed once
-   $ make run
-
-This will start a server that is exposed on port ``8000`` so now you can
-reach ``http://localhost:8000`` with your browser or curl.
-
-It will also start the ``create-react-app`` dev server on port ``3000``. That's
-the main URL to use.
-
-Lastly, you need to start the SQS daemon. This is started with:
-
-.. code-block:: shell
-
-   $ make daemon
-
-This will run until an unexpected error happens or until you kill it with ``Ctrl-C``.
 
 SQS Functional testing
 ======================

--- a/ui/package.json
+++ b/ui/package.json
@@ -24,5 +24,5 @@
     "lint:prettier": "prettier --list-different src/**/*.js | ./lint_problems.sh",
     "lint:prettierfix": "prettier src/**/*.js --write"
   },
-  "proxy": "http://localhost:8000"
+  "proxy": "http://web:8000"
 }

